validation/gas_token.go

```diff
@@ -0,0 +1,104 @@
+package validation
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"strings"
+	"testing"
+
+	"github.com/ethereum-optimism/superchain-registry/superchain"
+	"github.com/ethereum/go-ethereum"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/crypto"
+	"github.com/ethereum/go-ethereum/ethclient"
+	"github.com/stretchr/testify/require"
+)
+
+func testGasToken(t *testing.T, chain *superchain.ChainConfig) {
+	l1Client, err := ethclient.Dial(superchain.Superchains[chain.Superchain].Config.L1.PublicRPC)
+	require.NoError(t, err, "Failed to connect to the Ethereum client at RPC url %s", chain.PublicRPC)
+	defer l1Client.Close()
+
+	err = CheckGasToken(chain, l1Client)
+	require.NoError(t, err)
+}
+
+func CheckGasToken(chain *superchain.ChainConfig, l1Client EthClient) error {
+	l2Client, err := ethclient.Dial(chain.PublicRPC)
+	if err != nil {
+		return err
+	}
+	defer l2Client.Close()
+
+	weth9PredeployAddress := superchain.MustHexToAddress("0x4200000000000000000000000000000000000006")
+	want := "0000000000000000000000000000000000000000000000000000000000000020" + // offset
+		"000000000000000000000000000000000000000000000000000000000000000d" + // length
+		"5772617070656420457468657200000000000000000000000000000000000000" // "Wrapped Ether" padded to 32 bytes
+	gotName, err := getHexString("name()", weth9PredeployAddress, l2Client)
+	if err != nil {
+		return err
+	}
+	if want != gotName {
+		return fmt.Errorf("predeploy WETH9.name(): want=%s, got=%s", want, gotName)
+	}
+
+	l1BlockPredeployAddress := superchain.MustHexToAddress("0x4200000000000000000000000000000000000015")
+	isCustomGasToken, err := getBool("isCustomGasToken()", l1BlockPredeployAddress, l2Client)
+	if err != nil && !strings.Contains(err.Error(), "execution reverted") {
+		// Pre: reverting is acceptable
+		return err
+	} else {
+		// Post: must be set to false
+		if isCustomGasToken {
+			return fmt.Errorf("L1Block.isCustomGasToken() must return false")
+		}
+	}
+
+	isCustomGasToken, err = getBool("isCustomGasToken()", superchain.Addresses[chain.ChainID].SystemConfigProxy, l2Client)
```

Comment:
Shouldn't this use an l1Client?

I'd like us to revisit the "reverting is acceptable" case b/c it seems like it could be hiding many skeletons like this one. Can we double check it?

Reply:
Yea this should be an l1Client. Good catch. What are the pre/post comments referring to? Pre-what? Those comments were just copied from existing code

Reply:
I think what this is trying to do is to cope with deployments of the `SystemConfig` which don't expose the `isCustomGasToken()` method. This would indicate they don't support custom gas tokens and therefore should automatically pass the test. A better approach would be to read the semver here, and unless it is greater than or equal to some version (which we would need to look up or ask someone for) we don't do the rest of the check.

Reply:
Maybe there is a semver for the `SystemConfigProxy` we could check? 

If its < a certain value, we don't make the following calls at all (because the method doesn't exist?):
* SystemConfigProxy.isCustomGasToken()
* L1BlockPredeploy.isCustomGasToken()

Reply:
Discussed some findings on semver in [this discord thread](https://discord.com/channels/1244729134312198194/1319743226319474731/1319780018804756501). Not sure how to proceed at the moment. The `isCustomGasToken()` method seems to be missing on the latest standard version of the `SystemConfig` contract